test/e2e/node_feature_discovery_test.go

```diff
@@ -935,6 +963,158 @@ resyncPeriod: "1s"
+				return fmt.Errorf("node %q capacity does not match expected, diff (expected vs. received): %s", node.Name, cmp.Diff(expected, capacity))
+			}
+		}
+		return nil
+	}
+
+	return simplePoll(poll, 10)
+}
+
+// waitForNfdNodeAnnotations waits for node to be annotated as expected.
+func waitForNfdNodeAnnotations(ctx context.Context, cli clientset.Interface, expected map[string]string) error {
+	poll := func() error {
+		nodes, err := getNonControlPlaneNodes(ctx, cli)
+		if err != nil {
+			return err
+		}
+		for _, node := range nodes {
+			for k, v := range expected {
+				if diff := cmp.Diff(v, node.Annotations[k]); diff != "" {
+					return fmt.Errorf("node %q annotation does not match expected, diff (expected vs. received): %s", node.Name, diff)
+				}
+			}
+		}
+		return nil
+	}
+
+	return simplePoll(poll, 2)
+}
+
+type k8sLabels map[string]string
+
+// checkForNfdNodeLabels waits and checks that node is labeled as expected.
+func checkForNodeLabels(ctx context.Context, cli clientset.Interface, expectedNewLabels map[string]k8sLabels, oldNodes []corev1.Node) error {
+
+	poll := func() error {
+		nodes, err := getNonControlPlaneNodes(ctx, cli)
+		if err != nil {
+			return err
+		}
+		for _, node := range nodes {
+			nodeExpected, ok := expectedNewLabels[node.Name]
+			if !ok {
+				nodeExpected = k8sLabels{}
+				if defaultExpected, ok := expectedNewLabels["*"]; ok {
+					nodeExpected = defaultExpected
+				}
+			}
+
+			oldLabels := getNode(oldNodes, node.Name).Labels
+			expectedNewLabels := maps.Clone(oldLabels)
+			maps.Copy(expectedNewLabels, nodeExpected)
+
+			if !cmp.Equal(node.Labels, expectedNewLabels) {
+				return fmt.Errorf("node %q labels do not match expected, diff (expected vs. received): %s", node.Name, cmp.Diff(expectedNewLabels, node.Labels))
+			}
+		}
+		return nil
+	}
+
+	return simplePoll(poll, 3)
+}
+
+type k8sAnnotations map[string]string
+
+// checkForNodeAnnotations waits and checks that node is annotated as expected.
+func checkForNodeAnnotations(ctx context.Context, cli clientset.Interface, expectedNewAnnotations map[string]k8sAnnotations, oldNodes []corev1.Node) error {
+
+	poll := func() error {
+		nodes, err := getNonControlPlaneNodes(ctx, cli)
+		if err != nil {
+			return err
+		}
+		for _, node := range nodes {
+			nodeExpected, ok := expectedNewAnnotations[node.Name]
+			if !ok {
+				nodeExpected = k8sAnnotations{}
+				if defaultExpected, ok := expectedNewAnnotations["*"]; ok {
+					nodeExpected = defaultExpected
+				}
+			}
+
+			oldAnnotations := getNodeAnnotations(oldNodes, node.Name)
+			expectedNewAnnotations := maps.Clone(oldAnnotations)
+			maps.Copy(expectedNewAnnotations, nodeExpected)
+
+			if !cmp.Equal(node.Annotations, expectedNewAnnotations) {
+				return fmt.Errorf("node %q annotations do not match expected, diff (expected vs. received): %s", node.Name, cmp.Diff(expectedNewAnnotations, node.Annotations))
+			}
+		}
+		return nil
+	}
+
+	// Simple and stupid re-try loop
```

Comment:
We now have a more elegant way of running this loop, look at https://github.com/kubernetes-sigs/node-feature-discovery/pull/1184